### PR TITLE
Add operator for efficiently routing interesting values to downstream subscribers

### PR DESCRIPTION
**Pull request by lhotari**
## What?

This PR is about adding a stream operator that connects a source stream to downstream subscribers by providing a way to efficiently filter interesting values while keeping reactive stream semantics.

The working name has been RoutingFlux, but this might not be the best naming and could be changed.

## Why add a new stream operator for this purpose?

To provide an efficient way to filter interesting values while keeping reactive stream semantics. This can be used as a building block for different type of specialized use cases.

## State of the code in this PR

This is a very initial stage prototype. The code is not ready to be merged. It needs cleanup, polishing, refactoring and properly defining different test cases to cover all features that are part of the PR. The reason to make the PR early is to get early feedback.

### CODE REVIEW

**Comment by lhotari** [comment]
I was thinking that `AssertSubscriber` would be useful for writing some test cases. 
However the problem is that [`AssertSubscriber`](https://github.com/reactor/reactor-core/blob/master/src/test/java/reactor/test/subscriber/AssertSubscriber.java) currently exists in the `test` source set of `reactor-core` and isn't published in a library. 

@smaldini Is it intentional that `AssertSubscriber` isn't published? If it can be published, should it be moved to [`reactor-addons/reactor-test`](https://github.com/reactor/reactor-addons/tree/master/reactor-test/src/main/java/reactor/test) so that it can be used for writing the test cases for `RoutingFlux`?

**Comment by rajinisivaram** [comment]
@lhotari @smaldini I want to clarify the scope of `RoutingFlux`. This is not to suggest that the scope needs to be increased, but just to make sure that the limitations, particularly when consuming data from message brokers are clear.

Using Kafka as an example, consider the following scenarios:
1. Application processing latest values - durability not required, start with latest value on a topic, offsets are never committed, occasional message loss can be tolerated. This works well with `RoutingFlux`.
2. Messages routed to different subscribers based on Kafka partition (e.g. load balancing) - this already works well with Kafka consumer groups. `RoutingFlux` doesn't add any value.
3. Messages routed to different subscribers based on message key, messages from one partition may be split across different subscribers. Offset commit needs to take into account the processed offsets on all subscribers, making commits messy to implement.
4. Messages routed to multiple subscribers: Acknowledgements need to be reference counted so that offsets are committed only after all subscribers acknowledge, but only `RoutingFlux` has the count.

Perhaps we want to limit the scope of `RoutingFlux` to the first scenario. Or do more to help with acknowledgements/commits when used with message brokers. Thoughts?

**Comment by lhotari** [comment]
@rajinisivaram Thanks for the examples of possible use cases related to reactor-kafka. I'm sorry for the long delay in my reply.

> Perhaps we want to limit the scope of RoutingFlux to the first scenario. Or do more to help with acknowledgements/commits when used with message brokers. Thoughts?

I think this type of feedback is very helpful. I don't think it's a problem to extend the scope of the discussion to cover more advanced use cases. Optimally we would find a way to come up with a small core with minimal feature set for routing and some extended layer on top of it that also contains features related to message acknowledgement. I think that the processed Kafka offset tracking (scenario number 4) is a relevant use case. @rajinisivaram Do you have a proposal what the interface / API could look like?

**Comment by simonbasle** [comment]
@lhotari appart from the standing discussion around message acknowledgement that needs to happen with @rajinisivaram, what is the general state of this? Is it in an MVP state that we can further iterate upon later, or should we still defer merging for now?

**Comment by lhotari** [comment]
@simonbasle This PR is still in very early MVP state. I opened the PR to get some initial feedback. 
Could you or @smaldini reply [to the question about AssertSubscriber](https://github.com/reactor/reactor-addons/pull/103#issuecomment-293845447)? That would help me deciding how to start writing tests for the features in this PR.

**Comment by simonbasle** [comment]
yes this is intentional, you should prefer using `StepVerifier` from the `reactor-test` artifact in this repo (see http://projectreactor.io/docs/core/release/reference/docs/index.html#testing)

**Comment by lhotari** [comment]
> yes this is intentional, you should prefer using StepVerifier from the reactor-test artifact in this repo

Ok. I'll prefer StepVerifier. I remember some cases where AssertSubscriber would be easier to work with, but it maybe StepVerifier works as well.  

btw. I'm planning to check out whether https://github.com/reactor/reactor-core/pull/525 changes should affect the direction of the next iteration for this PR. The code in the initial version started off as [a copy of FluxPublish code](https://github.com/reactor/reactor-core/compare/master...lhotari:lh-routingflux-spike) so there might be now better ways to achieve what is being done. :) I'd appreciate guidance in showing the direction.

File: docs/src/docs/asciidoc/routing.adoc [hunk 1]
```diff
@@ -0,0 +1,83 @@
+[[routing]]
```
Comment (simonbasle) [inline docs/src/docs/asciidoc/routing.adoc]: Note that this folder has completely outdated content and is not linked or built when building the addons artifact. See #82. That work alone could be a reason to "revive" the asciidoc guide though...

**Comment by simonbasle** [comment]
@lhotari yes, `FluxPublish` has changed a bit in that commit so you should probably make similar changes **only for those that impact the `FluxPublish` class**. IIRC @smaldini did detect a potential bug in the previous implementation of FluxPublish.

**Pull request closed without merging**